### PR TITLE
Crossing out issues

### LINKED ISSUE #153: Cross out resolved issues
When first submitting a PR and getting this:
  
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%">2 Warnings</th>
     </tr>
  </thead>
  <tbody>
        <tr>
      <td>:warning:</td>
      <td>I don't like the code style</td>
    </tr>
            <tr>
      <td>:warning:</td>
      <td>Too many lines of code</td>
    </tr>
    
  </tbody>
</table>

Let's assume I'd push another commit fixing the issue, we could run a diff and do the following

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%">2 Warnings</th>
     </tr>
  </thead>
  <tbody>
        <tr>
      <td>:white_check_mark: </td>
      <td><strike>I don't like the code style</strike></td>
    </tr>
            <tr>
      <td>:warning:</td>
      <td>Too many lines of code</td>
    </tr>
    
  </tbody>
</table>

Idea from @camacho :+1:

### PR DESCRIPTION
**Pull request by marcelofabri**
Will fix #153. 

- [x] Add (optional) `sticky` parameter to DSL methods
- [x] Refactor warnings/messages/errors to be arrays of `Violation`
- [x] Add `data-sticky` to the table HTML
- [x] Make HTML tests green again
- [x] Read all violations from a HTML comment
- [x] Find out which ones should be crossed
- [x] Update comment with the crossed violations
- [x] Changelog
- [x] Tests 
- [x] Update README

### CODE REVIEW

**Comment by marcelofabri** [comment]
I don't know why Danger is warning about the gemspec, since I haven't touched it :grimacing:

**Comment by marcelofabri** [comment]
It seems that https://github.com/danger/danger/blob/master/Dangerfile#L10 should check if the array is not empty.

File: lib/danger/comment_generators/github.md.erb [hunk 1]
```diff
@@ -8,10 +8,10 @@
      </tr>
   </thead>
   <tbody>
-    <% table[:content].each do |message| -%>
+    <% table[:content].each do |violation| -%>
     <tr>
       <td>:<%= table[:emoji] %>:</td>
-      <td><%= message %></td>
+      <td data-sticky="<%= violation.sticky %>"><%= violation.message %></td>
```
Comment (jeroenvisser101) [inline lib/danger/comment_generators/github.md.erb]: You might want to add this, *only* if it's sticky, also, that would make sure that all previously made comments by Danger would be not-sticky, which I think is desired.

Comment (marcelofabri) [inline lib/danger/comment_generators/github.md.erb]: I think it's better to always have it, so in most cases it'll be explicit whether it's sticky. But good point on making it work with older comments!

**Comment by marcelofabri** [comment]
I've set `sticky` to be `true` by default, but I'm open to suggestions here

**Comment by jeroenvisser101** [comment]
@marcelofabri Hmm, not sure. I think that there's more that will be sticky than that will be non-sticky, so I guess making it opt-out is fine!

File: lib/danger/violation.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+module Danger
+  class Violation
+    attr_accessor :message, :sticky
+
+    def initialize(message, sticky)
```
Comment (orta) [inline lib/danger/violation.rb]: nice idea

**Comment by orta** [comment]
WRT the default of sticky, I'm not sure either, happy to try it as the default. But I feel like when I think about all the Dangerfiles I've made so far, I'd want very few of those notices as sticky.

**Comment by marcelofabri** [comment]
So, any ideas on how to get the previous violations without parsing the HTML? :grimacing: 

I've found Nokogiri to parse the comment HTML, but I remember @KrauseFx fighting for removing it from fastlane in the past ¯\_(ツ)_/¯

**Comment by orta** [comment]
Aye, definitely not by using nokogiri :p - I think regexes may have to work, or some smart string splitting. This is not a feature big enough to warrant another dependency.

**Comment by marcelofabri** [comment]
Just for reference, some regexes that seem to work:

* `<th width="100%">(.*)<\/th>` to find the table title
* `<td data-sticky="(.*)">(.*)<\/td>` to find if it's sticky and the violation text

I think we can get away without nokogiri :tada:

**Comment by marcelofabri** [comment]
Also, it seems that Circle is failing for some unrelated reason: https://circleci.com/gh/danger/danger/288

**Comment by jeroenvisser101** [comment]
@marcelofabri It could be a option to have JSON-encoded data inside `<table>` instead, and parse that. I think that would make this all a lot easier, but it would mean that previously generated comments have to be dealt with, as they do not contain this data.

**Comment by marcelofabri** [comment]
@jeroenvisser101 I'm not sure how we'd do that.

However, it wasn't that hard to parse the comment with regexes. I've pushed an initial implementation in c0e75c32d43f0d3c5303f875a87932ac20070aa0. I'll add more tests later.

**Comment by jeroenvisser101** [comment]
> I'm not sure how we'd do that.

I'm guessing we could parse the HTML with nokogiri and parse it directly? But I heard that nokogiri isn't a wanted dependency?

**Comment by jeroenvisser101** [comment]
@marcelofabri Parsing looks good! Let's keep this and do the JSON thing whenever we feel like we need it :)

**Comment by marcelofabri** [comment]
I need opinions here: should I leave the number of violations on the table header count the remaining violations or the total (including the crossed out ones)?

**Comment by jeroenvisser101** [comment]
@marcelofabri – I'd saying remaining

**Comment by marcelofabri** [comment]
@jeroenvisser101 wouldn't be weird to have "0 Errors"?

**Comment by jeroenvisser101** [comment]
We could make a check for zero and say something like ':white_check_mark: All good'

**Comment by marcelofabri** [comment]
@jeroenvisser101 I liked that! :100:

**Comment by KrauseFx** [comment]
@jeroenvisser101 No `nokogiri` please - @orta, @alloy and @supermarin wouldn't like this dependency :+1:

**Comment by jeroenvisser101** [comment]
> No nokogiri please

@KrauseFx – Alright! Any reason though? (I'm about to use it for another project and am wondering if I should consider to avoid it)


> I liked that!

@marcelofabri – Nice!

File: lib/danger/dangerfile_dsl.rb [hunk 3]
```diff
@@ -42,7 +44,9 @@ def fail(message, sticky = true)
       #
       # @param    [String] message
       #           The message to present to the user
-      def warn(message, sticky = true)
+      # @param    [Boolean] sticky
+      #           Whether the message should be kept after it was fixed
+      def warn(message, sticky: true)
```
Comment (jeroenvisser101) [inline lib/danger/dangerfile_dsl.rb]: We don't support ruby <2.0 right? Otherwise this could be a blocker

File: lib/danger/request_sources/github.rb [hunk 4]
```diff
@@ -174,12 +192,12 @@ def parse_comment(comment)
...
 
-      violations
+      violations.reject { |k, v| v.empty? }
     end
 
     def violations_from_table(table)
-      regex = %r{<td data-sticky="true">(<strike>)?(.*?)(</strike>)?</td>}im
-      table.scan(regex).flatten.compact.map(&:strip)
+      regex = %r{<td data-sticky="true">(<del>)?(.*?)(</del>)?\s*</td>}im
+      table.scan(regex).map { |a| a[1] }.map(&:strip)
```
Comment (jeroenvisser101) [inline lib/danger/request_sources/github.rb]: If we're only using the second match, we could simplify this (without 2 maps) by using non-capturing groups like this:

`<td data-sticky="true">(?:<del>)?(.*?)(?:</del>)?\s*</td>`

File: lib/danger/request_sources/github.rb [hunk 5]
```diff
@@ -174,12 +192,12 @@ def parse_comment(comment)
         violations[kind] = violations_from_table(table)
       end
 
-      violations
+      violations.reject { |k, v| v.empty? }
```
Comment (jeroenvisser101) [inline lib/danger/request_sources/github.rb]: Can we leave `k` here? Or if not, rename to `_` to indicate it's not being used?

File: lib/danger/dangerfile_dsl.rb (hunk 3 above)
Comment (marcelofabri) [inline lib/danger/dangerfile_dsl.rb]: I don't think so, because there're already some named parameters on the codebase

Comment (jeroenvisser101) [inline lib/danger/dangerfile_dsl.rb]: Alright, then this is totally cool :)

**Comment by KrauseFx** [comment]
Regarding nokogiri: Mostly because it might cause installation issues on Mac computers. It was the number one installation issue people experienced

File: lib/danger/request_sources/github.rb (hunk 5 above)
Comment (marcelofabri) [inline lib/danger/request_sources/github.rb]: :+1:

File: lib/danger/request_sources/github.rb (hunk 4 above)
Comment (marcelofabri) [inline lib/danger/request_sources/github.rb]: I didn't know about non-capturing groups. Nice!

**Comment by jeroenvisser101** [comment]
@KrauseFx ah, because of the native extensions?

File: lib/danger/request_sources/github.rb [hunk 6]
```diff
@@ -192,12 +192,12 @@ def parse_comment(comment)
...
 
-      violations.reject { |k, v| v.empty? }
+      violations.reject { |_, v| v.empty? }
     end
 
     def violations_from_table(table)
-      regex = %r{<td data-sticky="true">(<del>)?(.*?)(</del>)?\s*</td>}im
-      table.scan(regex).map { |a| a[1] }.map(&:strip)
+      regex = %r{<td data-sticky="true">(?:<del>)?(.*?)(?:</del>)?\s*</td>}im
+      table.scan(regex).flatten.map(&:strip)
```
Comment (jeroenvisser101) [inline lib/danger/request_sources/github.rb]: Looks good! :tada:

**Comment by jeroenvisser101** [comment]
Should we have a look and see why CircleCI is failing (even if unrelated?)

File: lib/danger/request_sources/github.rb (hunk 4 above)
Comment (jeroenvisser101) [inline lib/danger/request_sources/github.rb]: They can be really useful, sometimes :)

**Comment by KrauseFx** [comment]
We've had this issue with fastlane too, but can't remember how we fixed it. Maybe @mfurtak knows?

**Comment by marcelofabri** [comment]
This is ready for review :tada:

**Comment by mfurtak** [comment]
@KrauseFx I'll start a new PR to experiment a bit with build settings and report back if I find a solution :+1:

**Comment by mfurtak** [comment]
@KrauseFx #158 contains an adjustment that makes my branch pass.

**Comment by KrauseFx** [comment]
Thanks @mfurtak for fixing the build. @marcelofabri you should be able to merge master into your branch to make the build green :+1:

**Comment by marcelofabri** [comment]
Done! Thanks @mfurtak :100:

**Comment by orta** [comment]
Looks good :+1:
